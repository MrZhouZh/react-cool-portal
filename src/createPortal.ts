import { SFC, ReactPortal, useState, useEffect } from 'react';
import { createPortal } from 'react-dom';

import { RCPF, Props } from './types';

const createEl = (id: string): HTMLDivElement => {
  const el = document.createElement('div');
  el.setAttribute('id', id);
  document.body.appendChild(el);

  return el;
};

export default (
  id: string,
  isShow: boolean,
  hide: RCPF<MouseEvent> | false
): SFC<Props> => ({ children }: Props): ReactPortal => {
  const [container, setContainer] = useState(null);

  useEffect(() => {
    setContainer(document.getElementById(id) || createEl(id));

    return (): void => {
      if (!container) return;

      const timer = setTimeout(() => {
        clearTimeout(timer);
        if (container.innerHTML === '') container.remove();
      }, 100);
    };
  }, [container]);

  useEffect(() => {
    console.log('LOG ===> Portal rendered!');

    if (!hide || !container) return;

    const handler = (e: MouseEvent): void => {
      if (!container.contains(e.target)) hide(e);
    };

    document.addEventListener('click', handler);

    // eslint-disable-next-line consistent-return
    return (): void => {
      document.removeEventListener('click', handler);
    };
  }, [container]);

  return isShow && container && createPortal(children, container);
};
