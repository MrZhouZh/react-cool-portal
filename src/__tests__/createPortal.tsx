import React from "react";
import { render, fireEvent } from "@testing-library/react";

import createPortal from "../createPortal";
import { defaultContainerId } from "..";

describe("createPortal", () => {
  const childId = "test";

  interface Args {
    containerId?: string;
    isShow?: boolean;
  }
  interface Return {
    clickOutsideCb: any;
    escCb: any;
    baseElement: any;
    getByTestId: any;
    unmount: any;
  }

  const renderHelper = ({
    containerId = defaultContainerId,
    isShow = true,
  }: Args = {}): Return => {
    const clickOutsideCb = jest.fn();
    const escCb = jest.fn();
    const Portal = createPortal(containerId, isShow, clickOutsideCb, escCb);
    const { baseElement, getByTestId, unmount } = render(
      <Portal>
        <div data-testid={childId}>Test</div>
      </Portal>
    );

    return { clickOutsideCb, escCb, baseElement, getByTestId, unmount };
  };

  it("should not render portal", () => {
    const { baseElement } = renderHelper({ isShow: false });
    expect(baseElement).toMatchSnapshot();
  });

  it("should render portal", () => {
    const { baseElement } = renderHelper();
    expect(baseElement).toMatchSnapshot();
  });

  it("should auto remove container", () => {
    jest.useFakeTimers();

    const { baseElement, unmount } = renderHelper();
    unmount();
    jest.runAllTimers();
    expect(baseElement).toMatchSnapshot();
  });

  it("should trigger callback when clicks outside of the child", () => {
    const { clickOutsideCb } = renderHelper();
    fireEvent.click(document);
    expect(clickOutsideCb).toHaveBeenCalled();
  });

  it("should not trigger callback when clicks inside of the child", () => {
    const { getByTestId, clickOutsideCb } = renderHelper();
    fireEvent.click(getByTestId(childId));
    expect(clickOutsideCb).not.toHaveBeenCalled();
  });

  it("should trigger callback when presses ESC key", () => {
    const { escCb } = renderHelper();
    fireEvent.keyDown(document, { keyCode: 27 });
    expect(escCb).toHaveBeenCalled();
  });

  it("should not trigger callbacks when isShow set to false", () => {
    const { clickOutsideCb, escCb } = renderHelper({ isShow: false });
    fireEvent.click(document);
    fireEvent.keyDown(document, { keyCode: 27 });
    expect(clickOutsideCb).not.toHaveBeenCalled();
    expect(escCb).not.toHaveBeenCalled();
  });
});
