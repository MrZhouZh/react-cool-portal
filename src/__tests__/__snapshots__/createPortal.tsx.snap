// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPortal should auto remove container 1`] = `
<body>
  <div />
</body>
`;

exports[`createPortal should not auto remove container 1`] = `
<body>
  <div />
  <div
    id="react-cool-portal"
  />
</body>
`;

exports[`createPortal should not render portal 1`] = `
<body>
  <div />
</body>
`;

exports[`createPortal should render portal 1`] = `
<body>
  <div />
  <div
    id="react-cool-portal"
  >
    <div
      data-testid="target"
    >
      Test
    </div>
  </div>
</body>
`;
